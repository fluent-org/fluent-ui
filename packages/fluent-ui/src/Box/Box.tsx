import * as React from 'react'
import classNames from 'classnames'
import { createUseStyles, withTheme } from '@fluent-ui/styles'
import propTypes from '@styled-system/prop-types'
import { styles } from './Box.styled'
import { BoxClassProps, BoxProps, BoxPropTypes } from './Box.type'
import { omit } from '../utils'
import { Theme } from '../styles'

export const name = 'Box'

const useStyles = createUseStyles<Theme, BoxClassProps>(styles, { name })

const Box: React.FC<BoxProps> = React.forwardRef<HTMLDivElement, BoxProps>(
  (props, ref): React.ReactElement => {
    const {
      as: Component = 'div',
      className: classNameProp,
      id,
      style,
      onClick,
      children,
      acrylic = false,
      ...rest
    } = props
    const classes = useStyles(props)
    const className = classNames(
      classes.root,
      classes.system,
      {
        [classes.acrylic]: acrylic
      },
      classNameProp
    )

    const otherProps = omit(rest, [
      'theme',
      'minHeight',
      'minWidth',
      'maxWidth',
      'maxHeight',
      'bg',
      'background',
      'backgroundClip',
      'backgroundColor',
      'backgroundImage',
      'backgroundOrigin',
      'backgroundPosition',
      'backgroundRepeat',
      'backgroundSize',
      'backgroundAttachment',
      'border',
      'boxShadow',
      'flex',
      'flexBasis',
      'flexDirection',
      'flexFlow',
      'flexGrow',
      'flexShrink',
      'flexWrap',
      'order',
      'alignContent',
      'alignItems',
      'alignSelf',
      'justifyContent',
      'placeContent',
      'rowGap',
      'columnGap',
      'gap',
      'padding',
      'margin'
    ])

    const _props = {
      className,
      ref,
      id,
      style,
      onClick,
      ...otherProps
    }

    return <Component {..._props}>{children}</Component>
  }
)

Box.displayName = `F${name}`

Box.propTypes = {
  ...propTypes.space,
  ...propTypes.color,
  ...propTypes.typography,
  ...propTypes.layout,
  ...propTypes.background,
  ...propTypes.border,
  ...propTypes.position,
  ...propTypes.shadow,
  ...propTypes.flexbox,
  ...propTypes.grid,
  ...BoxPropTypes
}

// @ts-ignore
export default withTheme(Box)
