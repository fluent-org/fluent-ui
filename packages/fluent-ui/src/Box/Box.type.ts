// https://www.smooth-code.com/open-source/xstyled/docs/system-props/
import * as CSS from 'csstype'
import * as React from 'react'
import * as PropTypes from 'prop-types'
import { StandardProps } from '..'

export interface ResponsiveType<T> {
  xs?: T
  sm?: T
  md?: T
  lg?: T
  xl?: T
}
export type SpaceBaseTypes = number | string | 'auto'
export type SpaceTypes = SpaceBaseTypes | ResponsiveType<SpaceBaseTypes>
export type LayoutTypes = SpaceTypes
export type GridTypes = boolean | 'auto' | ResponsiveType<number>

export interface StylesProps {
  /**
   * Space
   */
  margin?: SpaceTypes
  marginTop?: SpaceTypes
  marginRight?: SpaceTypes
  marginBottom?: SpaceTypes
  marginLeft?: SpaceTypes
  // alias for margin
  m?: SpaceTypes
  // alias for marginTop
  mt?: SpaceTypes
  // alias for marginRight
  mr?: SpaceTypes
  // alias for marginBottom
  mb?: SpaceTypes
  // alias for marginLeft
  ml?: SpaceTypes
  // alias for margin-left and margin-right
  mx?: SpaceTypes
  // alias for margin-top and margin-bottom
  my?: SpaceTypes
  padding?: SpaceTypes
  paddingTop?: SpaceTypes
  paddingRight?: SpaceTypes
  paddingBottop?: SpaceTypes
  paddingLeft?: SpaceTypes
  // alias for padding
  p?: SpaceTypes
  // alias for paddingTop
  pt?: SpaceTypes
  // alias for paddingRight
  pr?: SpaceTypes
  // alias for paddingBottom
  pb?: SpaceTypes
  // alias for paddingLeft
  pl?: SpaceTypes
  // alias for padding-left and padding-right
  px?: SpaceTypes
  // alias for padding-top and padding-bottom
  py?: SpaceTypes
  space?: SpaceTypes
  /**
   * Layout
   */
  display?: LayoutTypes
  width?: LayoutTypes
  height?: LayoutTypes
  maxWidth?: LayoutTypes
  minWidth?: LayoutTypes
  maxHeight?: LayoutTypes
  minHeight?: LayoutTypes
  size?: LayoutTypes
  verticalAlign?: CSS.VerticalAlignProperty<SpaceBaseTypes>
  layout?: CSS.DisplayProperty | LayoutTypes | CSS.VerticalAlignProperty<SpaceBaseTypes>
  /**
   * XGrids
   */
  row?: boolean
  col?: GridTypes
  /**
   * Typography
   */
  fontFamily?: CSS.FontFamilyProperty
  fontSize?:
    | CSS.FontSizeProperty<number | string>
    | ResponsiveType<CSS.FontSizeProperty<number | string>>
  fontWeight?: CSS.FontWeightProperty
  lineHeight?: CSS.LineHeightProperty<number>
  textAlign?: CSS.TextAlignProperty
  letterSpacing?: CSS.LetterSpacingProperty<string | number>
  color?: CSS.ColorProperty
  textTransform?: CSS.TextTransformProperty
  typography?: string
  /**
   * Flexboxes
   */
  alignItems?: CSS.AlignItemsProperty
  alignContent?: CSS.AlignContentProperty
  justifyContent?: CSS.JustifyContentProperty
  justifyItems?: CSS.JustifyItemsProperty
  flexWrap?: CSS.FlexWrapProperty
  flexBasis?: CSS.FlexBasisProperty<number>
  flexDirection?: CSS.FlexDirectionProperty
  flex?: CSS.FlexProperty<number>
  justifySelf?: CSS.JustifySelfProperty
  alignSelf?: CSS.AlignSelfProperty
  order?: number | CSS.Globals
  flexboxes?: string
  /**
   * Grids
   */
  gridGap?: CSS.GridGapProperty<number>
  gridColumnGap?: CSS.GridColumnGapProperty<'normal'>
  gridRowGap?: CSS.GridRowGapProperty<'normal'>
  gridColumn?: CSS.GridColumnProperty
  gridRow?: CSS.GridRowProperty
  gridAutoFlow?: CSS.GridAutoFlowProperty
  gridAutoColumns?: CSS.GridAutoColumnsProperty<number>
  gridAutoRows?: CSS.GridAutoRowsProperty<number>
  gridTemplateColumns?: CSS.GridTemplateColumnsProperty<'none' | number>
  gridTemplateRows?: CSS.GridTemplateRowsProperty<'none' | number>
  gridTemplateAreas?: CSS.GridTemplateAreasProperty
  gridArea?: CSS.GridAreaProperty
  grids?: string
  /**
   * Backgrounds
   */
  background?: CSS.BackgroundProperty<number>
  backgroundColor?: CSS.BackgroundColorProperty
  backgroundImage?: CSS.BackgroundImageProperty
  backgroundSize?: CSS.BackgroundSizeProperty<number>
  backgroundPosition?: CSS.BackgroundPositionProperty<number>
  backgroundRepeat?: CSS.BackgroundRepeatProperty
  backgrounds?: string
  /**
   * Positioning
   */
  position?: CSS.PositionProperty
  zIndex?: CSS.ZIndexProperty
  top?: CSS.TopProperty<number>
  right?: CSS.TopProperty<number>
  bottom?: CSS.TopProperty<number>
  left?: CSS.TopProperty<number>
  positioning?: string
  /**
   * Borders
   */
  border?: CSS.BorderProperty<'inherit'>
  borderTop?: CSS.BorderTopProperty<number>
  borderTopColor?: CSS.BorderTopColorProperty
  borderRight?: CSS.BorderRightProperty<number>
  borderRightColor?: CSS.BorderRightProperty<number>
  borderBottom?: CSS.BorderBottomProperty<number>
  borderBottomColor?: CSS.BorderBottomProperty<number>
  borderLeft?: CSS.BorderLeftProperty<number>
  borderLeftColor?: CSS.BorderLeftProperty<number>
  borderColor?: CSS.BorderColorProperty
  borderWidth?: CSS.BorderWidthProperty<number>
  borderStyle?: CSS.BorderStyleProperty
  borderRadius?: CSS.BorderRadiusProperty<number>
  boxShadow?: CSS.BoxShadowProperty
  borders?: string
  /**
   * Basics
   */
  opacity?: LayoutTypes
  overflow?: CSS.OverflowProperty
  transition?: CSS.TransitionProperty
  basics?: string
}

export interface BoxProps extends StylesProps, StandardProps {
  acrylic?: boolean
  children?: React.ReactNode
  as?: keyof JSX.IntrinsicElements
}

export const BoxPropTypes = {
  acrylic: PropTypes.bool,
  children: PropTypes.node,
  as: PropTypes.oneOf<
    | 'symbol'
    | 'object'
    | 'style'
    | 'title'
    | 'time'
    | 'link'
    | 'menu'
    | 'dialog'
    | 'text'
    | 'p'
    | 'col'
    | 'a'
    | 'abbr'
    | 'address'
    | 'area'
    | 'article'
    | 'aside'
    | 'audio'
    | 'b'
    | 'base'
    | 'bdi'
    | 'bdo'
    | 'big'
    | 'blockquote'
    | 'body'
    | 'br'
    | 'button'
    | 'canvas'
    | 'caption'
    | 'cite'
    | 'code'
    | 'colgroup'
    | 'data'
    | 'datalist'
    | 'dd'
    | 'del'
    | 'details'
    | 'dfn'
    | 'div'
    | 'dl'
    | 'dt'
    | 'em'
    | 'embed'
    | 'fieldset'
    | 'figcaption'
    | 'figure'
    | 'footer'
    | 'form'
    | 'h1'
    | 'h2'
    | 'h3'
    | 'h4'
    | 'h5'
    | 'h6'
    | 'head'
    | 'header'
    | 'hgroup'
    | 'hr'
    | 'html'
    | 'i'
    | 'iframe'
    | 'img'
    | 'input'
    | 'ins'
    | 'kbd'
    | 'keygen'
    | 'label'
    | 'legend'
    | 'li'
    | 'main'
    | 'map'
    | 'mark'
    | 'menuitem'
    | 'meta'
    | 'meter'
    | 'nav'
    | 'noindex'
    | 'noscript'
    | 'ol'
    | 'optgroup'
    | 'option'
    | 'output'
    | 'param'
    | 'picture'
    | 'pre'
    | 'progress'
    | 'q'
    | 'rp'
    | 'rt'
    | 'ruby'
    | 's'
    | 'samp'
    | 'script'
    | 'section'
    | 'select'
    | 'small'
    | 'source'
    | 'span'
    | 'strong'
    | 'sub'
    | 'summary'
    | 'sup'
    | 'table'
    | 'template'
    | 'tbody'
    | 'td'
    | 'textarea'
    | 'tfoot'
    | 'th'
    | 'thead'
    | 'tr'
    | 'track'
    | 'u'
    | 'ul'
    | 'var'
    | 'video'
    | 'wbr'
    | 'webview'
    | 'svg'
    | 'animate'
    | 'animateMotion'
    | 'animateTransform'
    | 'circle'
    | 'clipPath'
    | 'defs'
    | 'desc'
    | 'ellipse'
    | 'feBlend'
    | 'feColorMatrix'
    | 'feComponentTransfer'
    | 'feComposite'
    | 'feConvolveMatrix'
    | 'feDiffuseLighting'
    | 'feDisplacementMap'
    | 'feDistantLight'
    | 'feDropShadow'
    | 'feFlood'
    | 'feFuncA'
    | 'feFuncB'
    | 'feFuncG'
    | 'feFuncR'
    | 'feGaussianBlur'
    | 'feImage'
    | 'feMerge'
    | 'feMergeNode'
    | 'feMorphology'
    | 'feOffset'
    | 'fePointLight'
    | 'feSpecularLighting'
    | 'feSpotLight'
    | 'feTile'
    | 'feTurbulence'
    | 'filter'
    | 'foreignObject'
    | 'g'
    | 'image'
    | 'line'
    | 'linearGradient'
    | 'marker'
    | 'mask'
    | 'metadata'
    | 'mpath'
    | 'path'
    | 'pattern'
    | 'polygon'
    | 'polyline'
    | 'radialGradient'
    | 'rect'
    | 'stop'
    | 'switch'
    | 'textPath'
    | 'tspan'
    | 'use'
    | 'view'
  >([
    'symbol',
    'object',
    'style',
    'title',
    'time',
    'link',
    'menu',
    'dialog',
    'text',
    'p',
    'col',
    'a',
    'abbr',
    'address',
    'area',
    'article',
    'aside',
    'audio',
    'b',
    'base',
    'bdi',
    'bdo',
    'big',
    'blockquote',
    'body',
    'br',
    'button',
    'canvas',
    'caption',
    'cite',
    'code',
    'colgroup',
    'data',
    'datalist',
    'dd',
    'del',
    'details',
    'dfn',
    'div',
    'dl',
    'dt',
    'em',
    'embed',
    'fieldset',
    'figcaption',
    'figure',
    'footer',
    'form',
    'h1',
    'h2',
    'h3',
    'h4',
    'h5',
    'h6',
    'head',
    'header',
    'hgroup',
    'hr',
    'html',
    'i',
    'iframe',
    'img',
    'input',
    'ins',
    'kbd',
    'keygen',
    'label',
    'legend',
    'li',
    'main',
    'map',
    'mark',
    'menuitem',
    'meta',
    'meter',
    'nav',
    'noindex',
    'noscript',
    'ol',
    'optgroup',
    'option',
    'output',
    'param',
    'picture',
    'pre',
    'progress',
    'q',
    'rp',
    'rt',
    'ruby',
    's',
    'samp',
    'script',
    'section',
    'select',
    'small',
    'source',
    'span',
    'strong',
    'sub',
    'summary',
    'sup',
    'table',
    'template',
    'tbody',
    'td',
    'textarea',
    'tfoot',
    'th',
    'thead',
    'tr',
    'track',
    'u',
    'ul',
    'var',
    'video',
    'wbr',
    'webview',
    'svg',
    'animate',
    'animateMotion',
    'animateTransform',
    'circle',
    'clipPath',
    'defs',
    'desc',
    'ellipse',
    'feBlend',
    'feColorMatrix',
    'feComponentTransfer',
    'feComposite',
    'feConvolveMatrix',
    'feDiffuseLighting',
    'feDisplacementMap',
    'feDistantLight',
    'feDropShadow',
    'feFlood',
    'feFuncA',
    'feFuncB',
    'feFuncG',
    'feFuncR',
    'feGaussianBlur',
    'feImage',
    'feMerge',
    'feMergeNode',
    'feMorphology',
    'feOffset',
    'fePointLight',
    'feSpecularLighting',
    'feSpotLight',
    'feTile',
    'feTurbulence',
    'filter',
    'foreignObject',
    'g',
    'image',
    'line',
    'linearGradient',
    'marker',
    'mask',
    'metadata',
    'mpath',
    'path',
    'pattern',
    'polygon',
    'polyline',
    'radialGradient',
    'rect',
    'stop',
    'switch',
    'textPath',
    'tspan',
    'use',
    'view'
  ])
}
