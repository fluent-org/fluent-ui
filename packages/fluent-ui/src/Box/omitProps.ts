export const omitProps = [
  'theme',
  'display',
  'overflow',
  'width',
  'height',
  'minHeight',
  'minWidth',
  'maxWidth',
  'maxHeight',
  'fontSize',
  'fontFamily',
  'fontWeight',
  'color',
  'bg',
  'background',
  'backgroundClip',
  'backgroundColor',
  'backgroundImage',
  'backgroundOrigin',
  'backgroundPosition',
  'backgroundRepeat',
  'backgroundSize',
  'backgroundAttachment',
  'border',
  'borderTop',
  'borderRight',
  'borderBottom',
  'borderLeft',
  'borderTopWidth',
  'borderRightWidth',
  'borderBottomWidth',
  'borderLeftWidth',
  'borderTopStyle',
  'borderRightStyle',
  'borderBottomStyle',
  'borderLeftStyle',
  'borderTopColor',
  'borderRightColor',
  'borderBottomColor',
  'borderLeftColor',
  'borderRadius',
  'borderTopLeftRadius',
  'borderTopRightRadius',
  'borderBottomRightRadius',
  'borderBottomLeftRadius',
  'margin',
  'marginLeft',
  'marginRight',
  'marginTop',
  'marginBottom',
  'padding',
  'paddingLeft',
  'paddingRight',
  'paddingTop',
  'paddingBottom',
  'boxShadow',
  'flex',
  'flexBasis',
  'flexDirection',
  'flexFlow',
  'flexGrow',
  'flexShrink',
  'flexWrap',
  'order',
  'alignContent',
  'alignItems',
  'alignSelf',
  'justifyContent',
  'placeContent',
  'rowGap',
  'columnGap',
  'gap',
  'padding',
  'margin',
  'zIndex',
  'textAlign',
  'textShadow',
  'gridGap'
]
