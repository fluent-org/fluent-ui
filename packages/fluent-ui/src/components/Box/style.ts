// https://www.smooth-code.com/open-source/xstyled/docs/system-props/
import * as CSS from 'csstype'

export interface ResponsiveType<T> {
  xs?: T
  sm?: T
  md?: T
  lg?: T
  xl?: T
}
export type SpaceBaseTypes = number | string | 'auto'
export type SpaceTypes = SpaceBaseTypes | ResponsiveType<SpaceBaseTypes>
export type LayoutTypes = SpaceTypes
export type GridTypes = boolean | 'auto' | ResponsiveType<number>

export interface StylesProps {
  /**
   * Space
   */
  margin?: SpaceTypes
  marginTop?: SpaceTypes
  marginRight?: SpaceTypes
  marginBottom?: SpaceTypes
  marginLeft?: SpaceTypes
  // alias for margin
  m?: SpaceTypes
  // alias for marginTop
  mt?: SpaceTypes
  // alias for marginRight
  mr?: SpaceTypes
  // alias for marginBottom
  mb?: SpaceTypes
  // alias for marginLeft
  ml?: SpaceTypes
  // alias for margin-left and margin-right
  mx?: SpaceTypes
  // alias for margin-top and margin-bottom
  my?: SpaceTypes
  padding?: SpaceTypes
  paddingTop?: SpaceTypes
  paddingRight?: SpaceTypes
  paddingBottop?: SpaceTypes
  paddingLeft?: SpaceTypes
  // alias for padding
  p?: SpaceTypes
  // alias for paddingTop
  pt?: SpaceTypes
  // alias for paddingRight
  pr?: SpaceTypes
  // alias for paddingBottom
  pb?: SpaceTypes
  // alias for paddingLeft
  pl?: SpaceTypes
  // alias for padding-left and padding-right
  px?: SpaceTypes
  // alias for padding-top and padding-bottom
  py?: SpaceTypes
  space?: SpaceTypes
  /**
   * Layout
   */
  display?: CSS.DisplayProperty
  width?: LayoutTypes
  height?: LayoutTypes
  maxWidth?: LayoutTypes
  minWidth?: LayoutTypes
  maxHeight?: LayoutTypes
  minHeight?: LayoutTypes
  size?: LayoutTypes
  verticalAlign?: CSS.VerticalAlignProperty<SpaceBaseTypes>
  layout?:
    | CSS.DisplayProperty
    | LayoutTypes
    | CSS.VerticalAlignProperty<SpaceBaseTypes>
  /**
   * XGrids
   */
  row?: boolean
  col?: GridTypes
  /**
   * Typography
   */
  fontFamily?: CSS.FontFamilyProperty
  fontSize?: CSS.FontSizeProperty<number | string>
  fontWeight?: CSS.FontWeightProperty
  lineHeight?: CSS.LineHeightProperty<number>
  textAlign?: CSS.TextAlignProperty
  letterSpacing?: CSS.LetterSpacingProperty<string | number>
  color?: CSS.ColorProperty
  textTransform?: CSS.TextTransformProperty
  typography?: string
  /**
   * Flexboxes
   */
  alignItems?: CSS.AlignItemsProperty
  alignContent?: CSS.AlignContentProperty
  justifyContent?: CSS.JustifyContentProperty
  justifyItems?: CSS.JustifyItemsProperty
  flexWrap?: CSS.FlexWrapProperty
  flexBasis?: CSS.FlexBasisProperty<number>
  flexDirection?: CSS.FlexDirectionProperty
  flex?: CSS.FlexProperty<number>
  justifySelf?: CSS.JustifySelfProperty
  alignSelf?: CSS.AlignSelfProperty
  order?: number | CSS.Globals
  flexboxes?: string
  /**
   * Grids
   */
  gridGap?: CSS.GridGapProperty<number>
  gridColumnGap?: CSS.GridColumnGapProperty<'normal'>
  gridRowGap?: CSS.GridRowGapProperty<'normal'>
  gridColumn?: CSS.GridColumnProperty
  gridRow?: CSS.GridRowProperty
  gridAutoFlow?: CSS.GridAutoFlowProperty
  gridAutoColumns?: CSS.GridAutoColumnsProperty<number>
  gridAutoRows?: CSS.GridAutoRowsProperty<number>
  gridTemplateColumns?: CSS.GridTemplateColumnsProperty<'none' | number>
  gridTemplateRows?: CSS.GridTemplateRowsProperty<'none' | number>
  gridTemplateAreas?: CSS.GridTemplateAreasProperty
  gridArea?: CSS.GridAreaProperty
  grids?: string
  /**
   * Backgrounds
   */
  background?: CSS.BackgroundProperty<number>
  backgroundColor?: CSS.BackgroundColorProperty
  backgroundImage?: CSS.BackgroundImageProperty
  backgroundSize?: CSS.BackgroundSizeProperty<number>
  backgroundPosition?: CSS.BackgroundPositionProperty<number>
  backgroundRepeat?: CSS.BackgroundRepeatProperty
  backgrounds?: string
  /**
   * Positioning
   */
  position?: CSS.PositionProperty
  zIndex?: CSS.ZIndexProperty
  top?: CSS.TopProperty<number>
  right?: CSS.TopProperty<number>
  bottom?: CSS.TopProperty<number>
  left?: CSS.TopProperty<number>
  positioning?: string
  /**
   * Borders
   */
  border?: CSS.BorderProperty<'inherit'>
  borderTop?: CSS.BorderTopProperty<number>
  borderTopColor?: CSS.BorderTopColorProperty
  borderRight?: CSS.BorderRightProperty<number>
  borderRightColor?: CSS.BorderRightProperty<number>
  borderBottom?: CSS.BorderBottomProperty<number>
  borderBottomColor?: CSS.BorderBottomProperty<number>
  borderLeft?: CSS.BorderLeftProperty<number>
  borderLeftColor?: CSS.BorderLeftProperty<number>
  borderColor?: CSS.BorderColorProperty
  borderWidth?: CSS.BorderWidthProperty<number>
  borderStyle?: CSS.BorderStyleProperty
  borderRadius?: CSS.BorderRadiusProperty<number>
  boxShadow?: CSS.BoxShadowProperty
  borders?: string
  /**
   * Basics
   */
  opacity?: number
  overflow?: CSS.OverflowProperty
  transition?: CSS.TransitionProperty
  basics?: string
}
