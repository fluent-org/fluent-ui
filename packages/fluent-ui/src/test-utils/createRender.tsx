import * as React from 'react'
import * as PropTypes from 'prop-types'
import { render } from '@testing-library/react'
import { ThemeProvider } from '../ThemeProvider'

interface WrapperProps {
  children: React.ReactNode
}

const Wrapper: React.FC<WrapperProps> = ({ children }): React.ReactElement => (
  <ThemeProvider theme={{}}>{children}</ThemeProvider>
)

Wrapper.propTypes = {
  children: PropTypes.node.isRequired
}

const customRender = (ui: React.ReactElement, options: any) =>
  render(ui, { wrapper: Wrapper, ...options })

export { customRender as render }
