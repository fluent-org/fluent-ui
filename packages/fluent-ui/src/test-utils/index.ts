export * from '@testing-library/react'
// @ts-ignore
export * from './createRender'
