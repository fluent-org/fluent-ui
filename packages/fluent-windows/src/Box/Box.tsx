import * as React from 'react'
import classNames from 'classnames'
import { createUseStyles, withTheme } from '@fluent-windows/styles'
import propTypes from '@styled-system/prop-types'
import { styles } from './Box.styled'
import { BoxClassProps, BoxProps, BoxPropTypes } from './Box.type'
import { omit } from '../utils'
import { Theme } from '../styles'
import { omitProps } from './omitProps'

export const name = 'Box'

const useStyles = createUseStyles<Theme, BoxClassProps>(styles, { name })

const Box: React.FC<BoxProps> = React.forwardRef<HTMLDivElement, BoxProps>(
  (props, ref): React.ReactElement => {
    const {
      as: Component = 'div',
      className: classNameProp,
      id,
      style,
      onClick,
      children,
      acrylic = false,
      ...rest
    } = props
    const classes = useStyles(props)
    const className = classNames(
      classes.root,
      classes.system,
      {
        [classes.acrylic]: acrylic
      },
      classNameProp
    )

    const otherProps = omit(rest, omitProps)

    const _props = {
      className,
      ref,
      id,
      style,
      onClick,
      ...otherProps
    }

    return <Component {..._props}>{children}</Component>
  }
)

Box.displayName = `F${name}`

Box.propTypes = {
  ...propTypes.space,
  ...propTypes.color,
  ...propTypes.typography,
  ...propTypes.layout,
  ...propTypes.background,
  ...propTypes.border,
  ...propTypes.position,
  ...propTypes.shadow,
  ...propTypes.flexbox,
  ...propTypes.grid,
  ...BoxPropTypes
}

// @ts-ignore
export default withTheme(Box)
