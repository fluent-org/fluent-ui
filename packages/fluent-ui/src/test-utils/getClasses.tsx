import * as React from 'react'
import { createUseStyles } from '@fluent-ui/styles'
import { Styles } from 'jss'
import { render } from './createRender'

export function getClasses<T extends string>(styles: Styles, name: string): Record<T, string> {
  const useStyles = createUseStyles<T>(styles, { name })
  let classes: Record<T, string>
  function Util(props: any): React.ReactElement {
    classes = useStyles(props)
    return <div />
  }
  render(<Util />)
  // @ts-ignore
  return classes
}
