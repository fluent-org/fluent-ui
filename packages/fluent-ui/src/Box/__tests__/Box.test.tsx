import * as React from 'react'
import { render } from '../../test-utils'
import '@testing-library/jest-dom/extend-expect'

import Box from '../Box'

describe('Box', (): void => {
  const testText = 'Hello World'
  const testChildren = <span>{testText}</span>

  test('should be support ref', (): void => {
    const ref = React.createRef<HTMLDivElement>()
    const { container } = render(<Box ref={ref}>{testChildren}</Box>)
    const { current: element } = ref
    expect(element).toEqual(container.firstChild)
  })

  test('should render children', (): void => {
    const { container, getByText, sheets } = render(<Box>{testChildren}</Box>)
    expect(getByText(testText)).toBeInTheDocument()
    expect(container.firstChild).toMatchSnapshot()
    expect(sheets.toString()).toMatchSnapshot()
  })

  test('should be rendered as nav', (): void => {
    const { getByText } = render(<Box as="nav">{testText}</Box>)
    expect(getByText(testText).tagName).toStrictEqual('NAV')
    expect(getByText(testText)).toMatchSnapshot()
  })

  test('should be rendered with acrylic effects', (): void => {
    const { container } = render(<Box acrylic>{testChildren}</Box>)
    expect(container.firstChild).toMatchSnapshot()
  })
})
