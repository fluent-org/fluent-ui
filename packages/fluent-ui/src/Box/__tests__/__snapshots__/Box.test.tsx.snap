// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Box as 1`] = `
.c0 {
  -webkit-overflow-scrolling: touch;
}

@media (min-width:600px) {
  .c0::-webkit-scrollbar {
    width: 8px;
    background: transparent;
  }

  .c0::-webkit-scrollbar-thumb {
    border-radius: 8px;
    background-color: rgba(0,0,0,0.2);
  }

  .c0::-webkit-scrollbar-thumb:hover {
    background-color: rgba(0,0,0,0.4);
  }
}

<span
  className="c0"
>
  as span
</span>
`;

exports[`Box basic 1`] = `
.c0 {
  -webkit-overflow-scrolling: touch;
}

@media (min-width:600px) {
  .c0::-webkit-scrollbar {
    width: 8px;
    background: transparent;
  }

  .c0::-webkit-scrollbar-thumb {
    border-radius: 8px;
    background-color: rgba(0,0,0,0.2);
  }

  .c0::-webkit-scrollbar-thumb:hover {
    background-color: rgba(0,0,0,0.4);
  }
}

<div
  className="c0"
>
  basic
</div>
`;
